### PR TITLE
Move LightStateRTCState definition to .h file.

### PR DESCRIPTION
**Pull request by mmakaay**
# What does this implement/fix? 

The definition for struct LightStateRTCState is currently unavailable for use from custom light code, because it resides in `light_state.cpp`  instead of `light_state.h`.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

# Explain your changes

I am working on a light component, which implements a "disco mode". The idea is that I can push light state updates that are not saved, not published and that are executed immediately (instead of waiting for the next loop() for that light).

One feature related to this, is that after the disco mode is stopped, the light is restored to its state from before the disco state  changes. Easy enough if I can read the last known state from rtc memory, but the problem with that is that I cannot use the `LightStateRTCState` struct from my code.

A "solution" for now, is to define my own copy of the struct, and loading the data from rtc memory using that struct. This does work, but when the original `LightStateRTCState` gets modified, my copy won't get modified and restoring the last saved state from my code will likely start failing.

An alternative that I considered was calling `LightState::setup()`, since restore functionality is buried in there. This doesn't feel like the right solution to me though. The function of the setup function is, well... to setup the component. Not to update its state at random points during its lifetime.

BTW: The perfect solution would be when LightState provided `save()` and `restore()` as ready-to-use methods. That way it wouldn't even be a big issue when the struct definition wouldn't be living in an .h file. But that is something for a separate pull request I think.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by OttoWinter — CHANGES_REQUESTED** [review]
We should not expose this struct externally - as that should not be the concern of developers.

Could you elaborate on what you mean by this disco mode? If it's implemented as an effect, you can use `LightCall::set_save()` to configure whether the state change is saved, and override `start()` `stop()` to save/restore the light colors

**Comment by mmakaay** [comment]
Agreed, I would be happier when I would be able to restore the state via a simple API call. Saving the state is already taken care of. Restoring the last saved settings is currently fully integrated within the setup() method, and I don't want to call that one to just restore the state.

My work-around for this, was to copy the required bits of the restore code into my own component, for which I had to define my own compatible version of the struct to decode the stored state. Definitely not as clean as a simple restore() method, but the only way I saw fit to get this working.

Overriding `start()` and `stop()` in an effect might not be feasible. As far as I know, I cannot simply ask `LightState` for its current state to save it for later. All state is handled within the `LightCall` class, up to saving the new state to RTC mem. That is the only registration of the last saved state that I could find, which is why I went for restoring the original light state by reading the RTC mem info.

Note that I'm not implementing an effect here though, but a way to set the color of the light immediately through an action, without saving or publishing the state. I called it "disco", because that seems to be the general term for this  style of update, especially now Philips HUE also has a disco mode in the app. Further development could for example be to provide a websocket through which fast light updates can be pushed to the lamp.

I guess a disco mode for streaming color changes could be implemented through an effect, by letting some interface method set the next color to apply, and letting the effect read and apply that. This would take care of the no-save-no-publish bit. However, it would not take care of the set-color-immediately bit. Normally setting a color would wait till the next loop() for the light has arrived. My action pushes the color to the lamp right away, making the lamp respond as quickly as possible to light changes.

For a little demo of what I was able to do with this, check out [this community posting])https://community.home-assistant.io/t/hacking-yeelight-fw-enabling-lan-control/284406/202?u=mmakaay), where I provide some visual feedback during the OTA process. Since OTA blocks the main loop, the regular way of setting a light color would not work here. The disco mode update does work however, and allows me to change the lamp color to indicate that it is being flashed.

**TL;DR:**

* As far as I know, I cannot retrieve the last saved state / current state from a `LightState` object.
* There is no way to restore the last saved state from an API call.
* That's why I went for restoring the last state myself by reading the RTC mem and hence the request to have access to the required struct.
* I do agree that it is very much an implementation detail and not an API.

**So based on all this, what would you propose?**

Extend the LightState API to have a restore() method (my favorite)
and/or extend LightState with a method to retrieve the current and/or last saved state?

Or am I overlooking some options here to implement what I want?

**Comment by mmakaay** [comment]
Yes, close this one. I think I can implement what I need in a clean manner by making use of the `LighCall::from_light_color_values()`.

If I run into a wall with that idea, I might return with a PR to make storing and restoring light state easily accessible via a cleanly encapsulated API extension. But so far, the route from above looks promising.
